### PR TITLE
Fix headernt.c & -output-obj Unicode support

### PR DESCRIPTION
**Pull request by nojb**
This is a follow-up to #1200.  It does three things:

- Fixes a bad rebase of `read_runtime_path` of `headernt.c`
- Fixes the C signature of the `argv` argument in the C file generated by `-output-obj`
- Uses `WriteConsole` in `headernt.c` to properly print Unicode pathnames in error messages.

Together with #1357 this should fix the remaining Windows testsuite errors.

/cc @dra27

### CODE REVIEW

**Comment by nojb** [comment]
Rebased on `trunk`.

**Review by damiendoligez — APPROVED** [review]
LGTM. It's up to you if you want to wait for a review from @dra27, who is away until Monday.

**Comment by nojb** [comment]
Thanks! No problem waiting for David's review (only downside is that AppVeyor will remain broken until then).

**Comment by nojb** [comment]
I was waiting for @dra27's review, but I think we could/should merge this now since it fixes two bugs and several testsuite failures.  We can always revisit if necessary.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/26 07:39 +0000):
> I was waiting for @dra27's review, but I think we could/should merge
> this now since it fixes two bugs and several testsuite failures.  We
> can always revisit if necessary.

Agreed! I'm sure it can't make the situation worse it currently is so
let's go aheand and fix later what may have to be fixed.

**Comment by gasche** [comment]
On the other hand, (1) this stuff seems tricky and (2) it's a pain to fix-after-the-fact when there is a release branch. Damien (or someone else) may want to merge it now, but I'll personally wait to merge unless @dra27 signals that he is too busy with other things.

**Comment by dra27** [comment]
Sorry - fully back from vacation tomorrow (though I may have a chance to give this and #1363 a review later today)

File: bytecomp/bytelink.ml
Comment (alainfrisch) [inline bytecomp/bytelink.ml:490]: Should probably be renamed to char_os once the other PR is merged, right?

Comment (nojb) [reply bytecomp/bytelink.ml:490]: That's right, will update as necessary.

File: stdlib/headernt.c
Comment (alainfrisch) [inline stdlib/headernt.c]: Out of curiosity: do we actually need the 'W' suffix here?  I thought we were now compiling in a mode where unprefixed functions are aliases for *W ones.

Comment (nojb) [reply stdlib/headernt.c]: I think you are right.  Will fix.

File: stdlib/headernt.c
Comment (alainfrisch) [inline stdlib/headernt.c]: It's very unimportant, but I wonder if in principle it wouldn't be better to translate to UTF-8 here (it could be that the displayed filename cannot be represented in the current 8-bit codepage).  Or is the idea of an UTF-8 encoded text file crazy under Windows?

Comment (alainfrisch) [reply stdlib/headernt.c]: Sorry, I missed the fact that CP is indeed UTF-8 when OCaml is configured with Unicode support.

Comment (nojb) [reply stdlib/headernt.c]: Absolutely I think it makes good sense! I just did it this way to preserve existing behavior.

**Comment by nojb** [comment]
Wohoo, all AppVeyor tests pass! @shindere @dra27 

Travis failures seem unrelated.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2017/09/26 10:45 +0000):
> Wohoo, all AppVeyor tests pass! @shindere @dra27

So nice! Hope Inria's CI will be as happy once this gets merged!

Thank you so much for your hard work @nojb !

File: testsuite/tests/win-unicode/Makefile
Comment (dra27) [inline testsuite/tests/win-unicode/Makefile]: This seems a little convoluted to me - why not just create exec_tests.precheck with `exit 1` in it - the entry in .gitignore is then not necessary and lines 22-25 below can just be deleted?

Comment (dra27) [reply testsuite/tests/win-unicode/Makefile]: The `exec_tests.precheck` file could also then include a comment as to why the tests are disabled (stability - which I will look into, as I have an idea what it is...)

File: bytecomp/bytelink.ml
Comment (dra27) [reply bytecomp/bytelink.ml:490]: In the interests of getting at least one of our CI systems back up and running, can we merge this before #1363?

**Review by dra27 — CHANGES_REQUESTED** [review]
LGTM - cosmetic tweak to the exec_tests disabling only

**Comment by dra27** [comment]
Doesn't have to be done here - but are there tests we could add for the -output-obj failure and similarly for headernt producing garbage when passed Unicode?

**Comment by dra27** [comment]
I've just popped this branch into Inria CI's precheck

File: testsuite/tests/win-unicode/Makefile
Comment (nojb) [reply testsuite/tests/win-unicode/Makefile]: Fixed as suggested, thanks!

**Review by dra27 — APPROVED** [review]

**Comment by dra27** [comment]
Lovely, thanks - let's let AppVeyor run through and then merge and cherry-pick. The Travis failures are identical to the current ones on trunk, so I agree that they don't appear to be related.

**Comment by shindere** [comment]
David Allsopp (2017/09/27 01:34 -0700):
> I've just popped this branch into Inria CI's precheck

Thanks! Please let us now whether it works -- precheck was destroyed I
don't know why and I had to re-create it from scratch.

**Comment by shindere** [comment]
David Allsopp (2017/09/27 09:23 +0000):
> Lovely, thanks - let's let AppVeyor run through and then merge and
> cherry-pick. The Travis failures are identical to the current ones on
> trunk, so I agree that they don't appear to be related.

What's the problem exactly with Travis?

Can't have a look at it, sorry.

**Comment by nojb** [comment]
@shindere wrote:
> What's the problem exactly with Travis?

64-bit builds fail with dozens of errors like this one:
> ... testing 'fileio.ml': ocamlcFile "fileio.ml", line 1:
Error: Error on dynamically loaded library: /home/travis/build/ocaml/ocaml/testsuite/../otherlibs/unix/dllunix.so: /home/travis/build/ocaml/ocaml/testsuite/../otherlibs/unix/dllunix.so: undefined symbol: caml_atom_table

**Comment by gasche** [comment]
The errors all look similar to this one:

> Running tests from 'tests/lib-bigarray' ...
>  ... testing 'bigarrays.ml': ocamlcFile "bigarrays.ml", line 1:
>Error: Error on dynamically loaded library: /home/travis/build/ocaml/ocaml/testsuite/../otherlibs/unix/dllunix.so: /home/travis/build/ocaml/ocaml/testsuite/../otherlibs/unix/dllunix.so: undefined symbol: caml_atom_table

This is unrelated to the present PR, but it's a recent breakage. I'll send an email.

**Comment by shindere** [comment]
Thanks, @nojb. Trying to reproduce this here.

**Comment by nojb** [comment]
Thanks!

**Comment by dra27** [comment]
Cherry-picked to 4.06 in https://github.com/ocaml/ocaml/commit/36e0a3bdc514c750ae0d10981f21a27431a1efb2

**Pull request merged**
